extensions/amp-lightbox-viewer/0.1/amp-lightbox-viewer.js

```diff
@@ -369,6 +384,89 @@ export class AmpLightboxViewer extends AMP.BaseElement {
       this.previous_();
     }
   }
+
+  /**
+   * Display gallery view to show thumbnails of lightboxed elements
+   * @private
+   */
+  openGallery_() {
+    // Build gallery div for the first time
+    if (!this.gallery_) {
+      this.buildGallery_();
+    }
+    this.container_.setAttribute('gallery-view', '');
+  }
+
+  /**
+   * Close gallery view
+   * @private
+   */
+  closeGallery_() {
+    this.container_.removeAttribute('gallery-view');
+  }
+
+  /**
+   * Build lightbox gallery. This is called only once when user enter gallery
+   * view for the first time.
+   * @private
+   */
+  buildGallery_() {
+    // Build gallery
+    this.gallery_ = this.win.document.createElement('div');
+    this.gallery_.classList.add('-amp-lightbox-viewer-gallery');
+
+    // Initialize thumbnails
+    this.updateThumbnails_();
+
+    this.container_.appendChild(this.gallery_);
+
+    // Add go back button
+    const back = this.closeGallery_.bind(this);
+    this.buildButton_('Back', 'amp-lightbox-viewer-button-back', back);
+
+  }
+
+  /**
+   * Update thumbnails displayed in lightbox gallery.
+   * This function only supports initialization now.
+   * @private
+   */
+  updateThumbnails_() {
+    if (this.thumbnails_) {
+      // TODO: Need to update gallery if there's change with thumbnails
+      return;
+    }
+
+    // Initialize thumbnails from lightbox manager
+    this.thumbnails_ = [];
+    const thumbnailList = this.manager_.getThumbnails();
+    thumbnailList.forEach(thumbnail => {
```

Comment:
Well, whether it takes long time here or in vsync doesn't matter. What matters is whether that code has any blocking reads which it doesn't seem to have. Another approach would be to create all thumbnails first and then append them to DOM in vsync mutate.